### DISCUSSION TITLE
[MTR05]  week12 作業：JSON格式的資料該如何在 js 和 PHP 之間做正確接收與轉換

### DISCUSSION THREAD

**Discussion by chachachater** [Lidemy/forum#142] [其他問題] [answered]
### mattermost 帳號
selena，因為實在是在這邊卡好久，搞懂之後發現就是自己不太懂這四者之間的差異，附上個筆記供參考
[JSON String :　decode ? encode ? parse ? stringfy ?](https://hackmd.io/@ouR5x-oVSMy4d8R5uFsKNg/BJH-atmod)

### 作業名稱
W12 hw2

### 碰到的問題
在 Todo 的作業之中，我想要用 JavaScript 接收資料庫裡面存放的 JSON格式的資料，但即使做了 `json.parse()` 所得到的結果仍然是 string type，目前有把其他不相關的程式碼註解掉，也有把 JSON 丟到[JSONLint](https://jsonlint.com/)來測試JSON格式對不對，測試結果是正確的，但最後還是得到字串型式的資料，想請教原因或該往哪個方向思考

php 程式碼如下：
```php
<?php
  require_once('connection.php');
  header('Access-Control-Allow-Methods: GET');
  header('Access-Control-Allow-Origin: *');
  header('Content-type: application/json;charset=utf-8');

  $test = '[{"data_id":"0","content":"test0","is_done":0},{"data_id":"1","content":"test1","is_done":1}]';
  print_r(json_encode($test));
?>
```
javascript 程式碼如下：
```js
getData()
function getData() {
  const url = `http://localhost/dashboard/Todo/api_todo.php?token=a`
  const request = new XMLHttpRequest
  request.open('GET', url, true)
  request.addEventListener('load', () => {
    const response = request.response

    console.log(response)
    console.log(JSON.parse(request.response))

    return request.response
  })
  request.send()
}
```

最後結果如下：
![image](https://user-images.githubusercontent.com/71329979/121802822-07d98a00-cc71-11eb-94a0-d7b086073c57.png)

### 已解決，解法如下：

`$test = '[{"data_id":"0","content":"test0","is_done":0},{"data_id":"1","content":"test1","is_done":1}]';`，這邊的 `$test` 是字串型式，然後我在 php 檔案之中又做了一次`json_encode($test)`，所以就產生了回答裡面提到的 字串中的字串 的 response，修改成 `print_r($test);` 之後就可以正常 parse() 了

> json_encode (  ) :  Returns a **string** containing the JSON representation of the supplied value.  BY php 文件

![image](https://user-images.githubusercontent.com/71329979/121805088-2bee9880-cc7c-11eb-9552-ea865fb2c4eb.png)

總結:  如果本來就已經是 JSON String 的話 ，就不用再做一次 `json_encode()`  XD

**Comment by aszx87410** [comment, marked as answer]
你可以從 network tab 看到伺服器給了什麼 response，看那邊最準

我猜是字串中的字串，所以 parse 完之後還是字串

**Reply by chachachater** [reply]
查看之後結果是這樣↓，字串中的字串的意思我不太懂，是指 data_id 、 content、 is_done 這些 key ?
 還是指 key 裡面的 value@@?

`"[{\"data_id\":\"0\",\"content\":\"test0\",\"is_done\":0},{\"data_id\":\"1\",\"content\":\"test1\",\"is_done\":1}]"`

![image](https://user-images.githubusercontent.com/71329979/121803198-fc875e00-cc72-11eb-941f-4f41d1d69a06.png)

**Reply by aszx87410** [reply]
你的 response 最外層用 "" 包了起來
所以你的 response 大概就像是：

var response = `'"[…]"'`

這就是字串中的字串的意思，所以你 json parse 完當然還是一個字串

**Reply by chachachater** [reply]
沒錯，是這字串中的字串的問題QQ，解法和原因有更新上去了~thk u
